### PR TITLE
Add support multiple currencies with PostgreSQL composite type

### PR DESCRIPTION
**Pull request by Nitrino**
This PR includes:
- Adding `Money.Ecto.Composite.Type` to support multiple currencies
- Refactoring existing Ecto types for common naming style

### CODE REVIEW

File: lib/money/ecto_type.ex [hunk 1]
```diff
@@ -7,43 +7,46 @@ if Code.ensure_compiled?(Ecto.Type) do
     This type expects you to use a single currency.
     The currency must be defined in your configuration.
```
Comment (ffloyd) [inline lib/money/ecto_type.ex]: Maybe you should add deprecation note to moduledoc.

File: README.md [hunk 2]
```diff
@@ -35,6 +35,125 @@ Money.to_string(Money.new(1_234_56, :EUR), separator: ".", delimeter: ",", symbo
...
+```
+
+### Serialization to database with multiple currency
+`Money.Ecto.Composite.Type` Ecto type represents serialization of `Money.t` to [PostgreSQL Composite Types](https://www.postgresql.org/docs/11/rowtypes.html) with saving currency.
+
+1. Create migration with custom type:
+```elixir
+  def up do
+    execute """
+    CREATE TYPE public.money_with_currency AS (amount integer, currency char(3))
```
[the remark this answers is not in the document]
Comment (Nitrino) [reply README.md]: This is just an example, there is no goal to teach PostgreSQL.
Also, the library does not support cryptocurrency.

File: README.md [hunk 3]
```diff
@@ -35,6 +35,125 @@ Money.to_string(Money.new(1_234_56, :EUR), separator: ".", delimeter: ",", symbo
...
+    """
+  end
+
+  def down do
+    execute """
+    DROP TYPE public.money_with_currency
+    """
+  end
+```
```
[the remark this answers is not in the document]
Comment (Nitrino) [reply README.md]: I think in the future to add migration generators.

File: lib/money/ecto/amount_type.ex [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    def type, do: :integer
+
+    @spec cast(String.t | integer()) :: {:ok, Money.t}
+    def cast(val)
+    def cast(str) when is_binary(str) do
+      Money.parse(str)
+    end
+    def cast(int) when is_integer(int), do: {:ok, Money.new(int)}
+    def cast(%Money{} = money), do: {:ok, money}
+    def cast(%{"amount" => amount, "currency" => currency}),
```
[the remark this answers is not in the document]
Comment (Nitrino) [reply lib/money/ecto/amount_type.ex]: Good point. I will pay attention to it, now it is beyond the scope of the pull request.

File: lib/money/ecto/currency_type.ex [hunk 5]
```diff
@@ -0,0 +1,51 @@
+if Code.ensure_compiled?(Ecto.Type) do
+  defmodule Money.Ecto.Currency.Type do
```
[the remark this answers is not in the document]
Comment (Nitrino) [reply lib/money/ecto/currency_type.ex]: This Ecto type can be used with `Money.Ecto.Amount.Type` to store amount and currency in different fields of the database.
This is rather inconvenient, but I see no reason not to help with this.

File: lib/money/ecto/currency_type.ex [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    @moduledoc """
+    Provides a type for Ecto to store a currency.
+    The underlying data type is a string.
+
+    ## Migration Example
+
+        create table(:my_table) do
+          add :currency, :string
```
[the remark this answers is not in the document]
Comment (Nitrino) [reply lib/money/ecto/currency_type.ex]: Good catch 👍

File: lib/money/ecto_type.ex (hunk 1 above)
Comment (Nitrino) [reply lib/money/ecto_type.ex]: 👍

File: test/money/ecto/amount_type_test.exs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  test "cast/1 String" do
+    assert Type.cast("1000") == {:ok, Money.new(100000, :GBP)}
+    assert Type.cast("1234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("1,234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("£1234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("£1,234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("£ 1234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("£ 1,234.56") == {:ok, Money.new(123456, :GBP)}
+    assert Type.cast("£ 1234") == {:ok, Money.new(123400, :GBP)}
+    assert Type.cast("£ 1,234") == {:ok, Money.new(123400, :GBP)}
```
Comment (pdgonzalez872) [inline test/money/ecto/amount_type_test.exs]: Hi! I have a quick question about these tests and I guess how this would work in a real app. Say I have the following:

```
  1) test cast/1 String (Money.Ecto.Amount.TypeTest)
     test/money/ecto/amount_type_test.exs:20
     Assertion with == failed
     code:  assert Type.cast("R$ 1.234,00") == {:ok, Money.new(123400, :GBP)}
     left:  {:ok, %Money{currency: :GBP, amount: 123}}
     right: {:ok, %Money{currency: :GBP, amount: 123400}}
     stacktrace:
       test/money/ecto/amount_type_test.exs:30: (test)

```

Is this expected behavior? Thanks for creating the patch!

Comment (Nitrino) [reply test/money/ecto/amount_type_test.exs]: This behavior depends on your config.
You can configure the separator and delimiter. 
e.g.:
```elixir
config :money,
  default_currency: :GDP,
  separator: ".",
  delimeter: ","

iex(2)> Type.cast("R$ 1.234,00")
{:ok, %Money{amount: 123400, currency: :USD}}
```

If you do not configure these parameters, you will get the result as in your example.
```elixir
config :money,
  default_currency: :GDP,

iex(2)> Type.cast("R$ 1.234,00")
{:ok, %Money{amount: 123, currency: :GBP}}
```

I will think about the default values of these configurations, but this will most require a change to the major version.

But the casting of other currencies in the default I think should not occur.

Comment (pdgonzalez872) [reply test/money/ecto/amount_type_test.exs]: Thanks for the quick reply!

> But the casting of other currencies in the default I think should not occur

That's fair! But, that's not what happens. It tries to cast, and the result is then incorrect, right? Should we warn/tell the user somehow that this happened? I can imagine a little "surprise" to know that amounts are flowing through. Maybe raise an error? I'm unsure, and regardless, thank you for explaining + doing this work!

File: lib/money/ecto/currency_type.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (ffloyd) [reply lib/money/ecto/currency_type.ex]: Ok, I agree.

File: lib/money/ecto/amount_type.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (ffloyd) [reply lib/money/ecto/amount_type.ex]: It's a bit unclear for me why it's beyond the scope. But if workaround will be added in a near future - there is no problem :)

File: test/money/ecto/amount_type_test.exs (hunk 7 above)
Comment (ffloyd) [reply test/money/ecto/amount_type_test.exs]: > I can imagine a little "surprise" to know that amounts are flowing through.

Agree. But I think it's subject for separate issue/PR. May you open such issue and describe all related edge cases which you know?

Comment (Nitrino) [reply test/money/ecto/amount_type_test.exs]: I will fix cast in this PR.
Old Ecto type (deprecated `Money.Ecto.Type` ) will remain unchanged, and in this new Ecto type (`Money.Ecto.Amount.Type`) will be fixed.

Comment (pdgonzalez872) [reply test/money/ecto/amount_type_test.exs]: Completely agree with both of you. I don't know any edge cases other than the one I pointed out, almost by luck as I didn't know it was a bigger issue/beyond the scope of this patch. I'd love to help out in any way I can here, maybe an opportunity for me to contribute. But I think it would take longer to help/guide me than for you to tackle the problem. If you are ok with it taking longer than it would take you I'm happy to try! Are you on the Elixir Slack? If so, I'd love to ask more questions about this.

Here is the issue: https://github.com/liuggio/money/issues/98

Comment (Nitrino) [reply test/money/ecto/amount_type_test.exs]: Yes, I am in Elixir Sluck (nick: nitrino). You can write to me there.

Comment (Nitrino) [reply test/money/ecto/amount_type_test.exs]: Accept this PR in the current state and transfer the discussion to your issue.
